Feature/webp target size

john-parton opened the pull request:
This is the initial work to close https://github.com/libvips/libvips/issues/3824

It works, but there's one "gotcha." The `target_size` parameter doesn't do anything unless `pass` is 2 or greater. The more passes, the closer it will get to `target_size`. In my non-scientific tests, 4 to 6 seem to be reasonable numbers.

I did a basic 'smoke test' with the CLI tool:

```bash
vips rot myfile.jpg out.webp[Q=100,target_size=2500000,pass=4] d90
```

and python
```python
with open("myfile.jpg", "rb") as f:
    buffer = f.read()

image = pyvips.Image.new_from_buffer(buffer, "")

buff = image.webpsave_buffer(target_size=1_000_000, **{"pass": 4})

with open("out.webp", "wb") as f:
    f.write(buff)
```

Note that 'pass' is a keyword in python, so I have to pass it by unpacking a dictionary. Very not fun.

Does this general approach seem reasonable?

jcupitt reviewed (commented):
Looks good!

You'll need to also add a note to `ChangeLog` and a test to `test_webp` in `test/test_suite/test_foreign.py`.

jcupitt commented on libvips/foreign/webpsave.c:
```diff
@@ -132,6 +132,17 @@ typedef struct _VipsForeignSaveWebp {
+	 * If target_size is set, this must be at least 1 for target_size to
+	 * have an effect.
+	 */
+	int pass;
```
Maybe default this to 4 if `target_size` is set?

john-parton replied:
I think I would prefer `passes` or `analysis_passes`, but that's being picky.

I can add a default just to improve ergonomics, yeah. 

Notes: I found that `0` resulted in the feature not working at all. I stepped through the webp code and that's the intended result. Values of `1` and `2` resulted in the file not being close to the target size at all, at least to me subjectively, and values of `6` and higher resulted in it taking an usually long time to encode.

jcupitt commented on libvips/foreign/webpsave.c:
```diff
@@ -604,6 +615,10 @@ vips_foreign_save_webp_init_config(VipsForeignSaveWebp *webp)
 	webp->config.method = webp->effort;
+	webp->config.target_size = webp->target_size; 
+
+	if (webp->pass)
```
You could also use `if (vips_object_argument_isset(object, "pass")) {}`, it's a little safer.

john-parton replied:
Interesting. I'll read through the internals. This is my first contribution to this project, and I haven't seriously programmed in C for over a decade.

john-parton replied:
I don't see a variable called `object` in scope here. Could I have a little more guidance?

john-parton replied:
I think I figured out how to do it properly. Resolved.

jcupitt commented on libvips/foreign/webpsave.c:
```diff
@@ -1161,6 +1190,8 @@ vips_foreign_save_webp_mime_init(VipsForeignSaveWebpMime *mime)
  * * @alpha_q: %gint, set alpha quality in lossless mode
  * * @effort: %gint, level of CPU effort to reduce file size
+ * * @target_size: %gint, desired target size in bytes
+ * * @pass: %gint, number of entropy-analysis passes
```
And add a paragraph to the doc text below explaining how `target_size` works.

john-parton replied:
Let me know if I need to update the verbiage/phrasing.

john-parton commented on libvips/foreign/webpsave.c:
```diff
@@ -895,6 +910,20 @@ vips_foreign_save_webp_class_init(VipsForeignSaveWebpClass *class)
+
+	VIPS_ARG_INT(class, "pass", 23,
+		_("Passes"),
+		_("Number of entropy-analysis passes (in [1..10])"),
```
I think zero in this case is "default" which is actually "1". I'll have to step through the code again and verify.

john-parton commented:
An additional note: I'll have to verify, but I believe the target_size parameter doesn't *limit* the size, it just tries to make it "close" depending on the number of passes. I'll do some smoke tests.

john-parton commented:
I wanted to make sure that I pick a sane default for `passes`

Here's some python code I wrote to exercise this branch:

```python

TARGET_SIZE = 1_000_000

with open("myfile.jpg", "rb") as f:
    buffer = f.read()

image = pyvips.Image.new_from_buffer(buffer, "")

print("Default encoding")
with Bench("webpsave_buffer"):
    buff = image.webpsave_buffer()
print(f"Size: {len(buff)}")
print()

print()
for i in range(1, 11):
    with Bench(f"target_size={TARGET_SIZE} & passes={i}"):
        buff = image.webpsave_buffer(target_size=TARGET_SIZE, passes=i)
    print(
        f"Size with passes={i}: {len(buff)} ({100 * len(buff) / TARGET_SIZE:.2f}%)"
    )
    print()
```

And here's the output
```
Default encoding
Starting webpsave_buffer
webpsave_buffer completed, elapsed 0:00:01.005219
Size: 656040


Starting target_size=1000000 & passes=1
target_size=1000000 & passes=1 completed, elapsed 0:00:00.891990
Size with passes=1: 656040 (65.60%)

Starting target_size=1000000 & passes=2
target_size=1000000 & passes=2 completed, elapsed 0:00:01.725788
Size with passes=2: 1165504 (116.55%)

Starting target_size=1000000 & passes=3
target_size=1000000 & passes=3 completed, elapsed 0:00:02.498880
Size with passes=3: 1011412 (101.14%)

Starting target_size=1000000 & passes=4
target_size=1000000 & passes=4 completed, elapsed 0:00:03.260787
Size with passes=4: 1011438 (101.14%)

Starting target_size=1000000 & passes=5
target_size=1000000 & passes=5 completed, elapsed 0:00:03.260441
Size with passes=5: 1011438 (101.14%)

Starting target_size=1000000 & passes=6
target_size=1000000 & passes=6 completed, elapsed 0:00:03.263195
Size with passes=6: 1011438 (101.14%)

Starting target_size=1000000 & passes=7
target_size=1000000 & passes=7 completed, elapsed 0:00:03.270455
Size with passes=7: 1011438 (101.14%)

Starting target_size=1000000 & passes=8
target_size=1000000 & passes=8 completed, elapsed 0:00:03.263690
Size with passes=8: 1011438 (101.14%)

Starting target_size=1000000 & passes=9
target_size=1000000 & passes=9 completed, elapsed 0:00:03.262610
Size with passes=9: 1011438 (101.14%)

Starting target_size=1000000 & passes=10
target_size=1000000 & passes=10 completed, elapsed 0:00:03.261657
Size with passes=10: 1011438 (101.14%)
```

A few conclusions:

* In general, it doesn't limit the size to `target_size`, and it looks like even when it's "as close as possible", it's still a bit larger. I suspect that has to do with headers/container overhead.
* `passes == 1` makes the `target_size` parameter have no effect.
* `passes == 2` is off by over 15%
* `passes == 3` is off by only 1% and doesn't dramatically change encoding time.
* `passes == 4` is only 24 bytes different from `passes == 3`, but the encoding time is significantly longer.
* `passes > 4` has literally no effect at all with the test image I have.

john-parton commented:
I think this is ready for another review.

* Added a message in the ChangeLog
* Added description of parameters in doc comment
* Define a variable for default number of passes. Let me know if there's a preferred way of doing this. I didn't want to have a magic number just hard coded.
* Added a python test
* use the `isset` helper function

jcupitt commented on ChangeLog:
```diff
@@ -15,6 +15,8 @@ master 8.16
 - PFM save and load now uses scRGB (ie. linear 0-1) [NiHoel]
 - turn `vips_addalpha` into a VipsOperation [RiskoZoSlovenska]
 - add vips_rawsave_target(), vips_rawsave_buffer() [akash-akya]
+- webpsave has @target_size parameter to set desire size 
```
desired

jcupitt commented on libvips/foreign/webpsave.c:
```diff
@@ -84,6 +84,14 @@
 #include <webp/mux.h>
 
+/* By default, entropy passes is equal to one, but 
+ * if target_size is provided, this is a more reasonable
```
Should this be "if target_size is not provided"?

jcupitt commented on libvips/foreign/webpsave.c:
```diff
@@ -746,6 +767,9 @@ vips_foreign_save_webp_build(VipsObject *object)
 	}
 
+	if (!vips_object_argument_isset(object, "passes"))
+		webp->passes = vips_object_argument_isset(object, "target_size") ? TARGET_SIZE_DEFAULT_PASSES : 1;
```
You don't need the `if()` if you're using a ternary.

jcupitt replied:
ie. I'd write:

```C
if (!vips_object_argument_isset(object, "passes"))
		webp->passes = 3;
```

(I think a hardwired number is OK here)

john-parton replied:
I think there maybe is some misunderstanding here. `pass` _has_ an effect even if `target_size` isn't set. Changing it to be `3` here will cause an issue with making encoding slower for very little gain.

Increasing pass increases encoding time:

```
Starting passes=1
passes=1 completed, elapsed 0:00:01.128829
Size with passes=1: 656040

Starting passes=2
passes=2 completed, elapsed 0:00:01.599618
Size with passes=2: 655410

Starting passes=3
passes=3 completed, elapsed 0:00:02.295408
Size with passes=3: 655382

Starting passes=4
passes=4 completed, elapsed 0:00:03.015794
Size with passes=4: 655398

Starting passes=5
passes=5 completed, elapsed 0:00:03.729868
Size with passes=5: 655412

Starting passes=6
passes=6 completed, elapsed 0:00:04.411422
Size with passes=6: 655424

Starting passes=7
passes=7 completed, elapsed 0:00:05.138159
Size with passes=7: 655490

Starting passes=8
passes=8 completed, elapsed 0:00:05.855748
Size with passes=8: 655712

Starting passes=9
passes=9 completed, elapsed 0:00:06.545841
Size with passes=9: 655582

Starting passes=10
passes=10 completed, elapsed 0:00:07.225836
Size with passes=10: 655640
```

This is *without* a target_size set at all.

Exercise code:

```python3
with open("test.jpg", "rb") as f:
    buffer = f.read()

image = pyvips.Image.new_from_buffer(buffer, "")

for i in range(1, 11):
    with Bench(f"passes={i}"):
        buff = image.webpsave_buffer(passes=i)
    print(f"Size with passes={i}: {len(buff)}")
    print()
```

So if we set passes to three, it will increase encoding time for very minimal quality gains.

jcupitt commented on libvips/foreign/webpsave.c:
```diff
@@ -1183,7 +1223,11 @@ vips_foreign_save_webp_mime_init(VipsForeignSaveWebpMime *mime)
+ * 
+ * Use @target_size to set the desired target size in bytes.
  *
+ * Use @pass to set the number of entropy-analysis passes. It has the range 1-10.
```
I'd also explain the default, eg. "if you set @target_size and @pass is not set, @pass defaults to 3".

jcupitt commented on libvips/foreign/webpsave.c:
```diff
@@ -84,6 +84,14 @@
+ * default value.
+ */
+#ifndef TARGET_SIZE_DEFAULT_PASSES
+#define TARGET_SIZE_DEFAULT_PASSES 3
```
I don't mind a magic number for this. And it needs to be explained to the user in the docs, so it's not really a symbolic value.

jcupitt reviewed (approved):
Great!

jcupitt commented:
> I think I need help with code style. Is there an auto-formatter / linter that I can run locally?

Yes, run `git clang-format` after a change and it'll reformat the code you just entered. Or `clang-format -i xxxxx.c` to reformat one file.

john-parton commented:
I think this is ready for review again. I have a comment above, but I'll re-iterate it here.

It's important that `passes` _not_ be set to `3` if `target_size` _isn't_ set. Doing so will result in the encoder taking significantly longer for no gain.

kleisauke commented on libvips/foreign/webpsave.c:
```diff
@@ -746,6 +759,9 @@ vips_foreign_save_webp_build(VipsObject *object)
 		return -1;
 	}
 
+	if (!vips_object_argument_isset(object, "passes"))
+		webp->passes = vips_object_argument_isset(object, "target_size") ? 3 : 1;
```
`->passes` already defaults to `1`, so I think this could be written as:
```suggestion
	if (!vips_object_argument_isset(object, "passes") &&
		vips_object_argument_isset(object, "target_size"))
		webp->passes = 3;
```

jcupitt replied:
Yes, this sounds useful. You'll need a quick note in the docs explaining the behaviour as well.

john-parton commented:
I might need a little help documenting "passes".

So basically, it seems to only do something positive/beneficial when `target_size` or `target_PSNR` is set.

If you turn on passes without either target_size or target_PSNR, it just dramatically increases encoding time, while leaving both the size and PSNR of the resulting image essentially untouched.

Output:

```
Starting passes=1,Q=90
passes=1,Q=90 completed, elapsed 0:00:01.216998
Size with passes=1: 1776302

Starting passes=2,Q=90
passes=2,Q=90 completed, elapsed 0:00:01.913680
Size with passes=2: 1773404

Starting passes=3,Q=90
passes=3,Q=90 completed, elapsed 0:00:02.734768
Size with passes=3: 1773730

Starting passes=4,Q=90
passes=4,Q=90 completed, elapsed 0:00:03.536159
Size with passes=4: 1773784

Starting passes=5,Q=90
passes=5,Q=90 completed, elapsed 0:00:04.361937
Size with passes=5: 1773346

Starting passes=6,Q=90
passes=6,Q=90 completed, elapsed 0:00:05.169635
Size with passes=6: 1773392

Starting passes=7,Q=90
passes=7,Q=90 completed, elapsed 0:00:05.994716
Size with passes=7: 1773680

Starting passes=8,Q=90
passes=8,Q=90 completed, elapsed 0:00:06.808321
Size with passes=8: 1773834

Starting passes=9,Q=90
passes=9,Q=90 completed, elapsed 0:00:07.636810
Size with passes=9: 1773820

Starting passes=10,Q=90
passes=10,Q=90 completed, elapsed 0:00:08.462970
Size with passes=10: 1773620

PSNR with passes=1: 41.95693621828301

PSNR with passes=2: 41.95630049312027

PSNR with passes=3: 41.95636007882095

PSNR with passes=4: 41.95786021891513

PSNR with passes=5: 41.968248519712894

PSNR with passes=6: 41.956590374928

PSNR with passes=7: 41.96897815022609

PSNR with passes=8: 41.955997198410685

PSNR with passes=9: 41.956692915390896

PSNR with passes=10: 41.95625941346408
```

Exercise script:

```python3
with open("test.jpg", "rb") as f:
    buffer = f.read()

image = pyvips.Image.new_from_buffer(buffer, "")

for i in range(1, 11):
    with Bench(f"passes={i},Q={Q}"):
        buff = image.webpsave_buffer(passes=i, Q=Q)
    with open(f"out_{i}.webp", "wb") as f:
        f.write(buff)
    print(f"Size with passes={i}: {len(buff)}")
    print()

img_og = cv2.imread("test.jpg")

for i in range(1, 11):
    img = cv2.imread(f"out_{i}.webp")
    psnr = cv2.PSNR(img_og, img)
    print(f"PSNR with passes={i}: {psnr}")
    print()
```

I didn't implement `target_PSNR` because I didn't need it for my project, and it seems even more esoteric than `target_size`.

Not sure how to document this to basically say "this is a foot-gun if you don't have target_size set."

john-parton commented:
I think this if ready for a final review. I updated the docs to caution users against setting `passes` without setting `target_size`

john-parton commented:
@kleisauke Would you mind taking another look? I think this is pretty close.

jcupitt commented:
Ah I jumped the gun, there seem to be some settings issues.

With this branch, I see:

```
$ vips black x.webp 1000 1000
webpsave: invalid configuration
```

The tests above are failing with:

```
webpsave: unable to encode
webpsave: mux error
```

jcupitt commented:
Use `ninja test` to run those tests locally.

kleisauke commented:
Commit 8de3704f46927b5d329c0eded7de20f91b300551 should fix that.

kleisauke commented:
Ah, sorry, that only resolves `meson test`, the Python test suite seems to fail with:
```python
        # target_size should reasonably work, +/- 2% is fine
        x = pyvips.Image.new_from_file(WEBP_FILE)
        buf_size = len(x.webpsave_buffer(target_size=10_000))
>       assert 9800 < buf_size < 10200
E       assert 18684 < 10200
````

kleisauke commented:
Commit bedd618aaa0371c8528164e476aa30d7d44c162e should fix the Python test suite by increasing the `target_size` argument to `20000`, as the original image is already a WebP image of ~30 KiB.

I also passed `keep='none'` to strip all metadata, since that size is not included as part of the `target_size` calculations in libwebp.

john-parton commented:
Thanks for being so helpful and patient. This is my first time using meson/ninja and the first time I've touched C/C++ seriously in a while.

I merged in the fixes you wrote, squashed it down to one commit, and rebased on master to fix a merge conflict with the changelog.

Ran both `meson test` and `ninja test` this time.

jcupitt commented:
Great! This will be in 8.16, due in a few months.
